compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -61,6 +62,243 @@ static bool consumesInputOp(NamedSequenceOp op) {
+  std::string specBaseName = getBaseName(specName);
+  unsigned specNameSeenCount = specNameCounts[specBaseName]++;
+
+  if (specNameSeenCount == 0)
+    return specBaseName;
+
+  return llvm::formatv("{}_{}", specBaseName, specNameSeenCount).str();
+}
+
+struct TuningSpecsToMerge {
+  SmallVector<NamedSequenceOp> namedSequenceOpsToMove;
+  // Maps a `NamedSequenceOp` to its user `ForeachMatchOp`, in which the
+  // `NamedSequenceOp` is used as a matcher or action inside a
+  // `ForeachMatchOp`.
+  llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> namedSequenceToUser;
+};
+
+// Populates the mapping of `NamedSequenceOp` to `ForeachMatchOp`
+// by processing matchers and actions inside the given `ForeachMatchOp`.
+//
+// This function iterates over all matcher and action references in
+// `foreachMatch`. If a reference is a valid `SymbolRefAttr` that resolves to a
+// `NamedSequenceOp` within the `module`, the function records the association
+// by mapping the `NamedSequenceOp` to `foreachMatch` in `namedSequenceToUser`.
+static void updateForeachMatchMappings(
+    ForeachMatchOp foreachMatch, ModuleOp module,
+    llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> &namedSequenceToUser) {
+  for (auto matcher : foreachMatch.getMatchers()) {
+    if (auto matcherSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(matcher)) {
+      if (auto matcherOp =
+              SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+                  module, matcherSymRef)) {
+        namedSequenceToUser[matcherOp] = foreachMatch;
+      }
+    }
+  }
+
+  for (auto action : foreachMatch.getActions()) {
+    if (auto actionSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(action)) {
+      if (auto actionOp = SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+              module, actionSymRef)) {
+        namedSequenceToUser[actionOp] = foreachMatch;
+      }
+    }
+  }
+}
+
+// Collects tuning specs (`NamedSequenceOp`s) from all inner modules within the
+// given input model, and organizes them into a `TuningSpecsToMerge` struct.
+// - `namedSequenceOpsToMove`: Contains `NamedSequenceOp`s that either:
+//     - Are not named `__kernel_config`.
+//     - Do not have the `iree_codegen.tuning_spec_entrypoint` attribute.
+// - `namedSequenceToForeachMatch`: Maps `NamedSequenceOp`s used as matchers or
+//    actions to their corresponding `ForeachMatchOp`.
+static TuningSpecsToMerge collectTuningSpecsToMerge(ModuleOp module) {
+  TuningSpecsToMerge tuningSpecs;
+  for (auto innerModule : module.getBody()->getOps<ModuleOp>()) {
+    for (auto namedSequenceOp :
+         innerModule.getBody()->getOps<NamedSequenceOp>()) {
+      if (namedSequenceOp.getSymName() != kKernelConfigSpecName ||
+          !namedSequenceOp->hasAttr(kTuningSpecEntrypointAttrName)) {
+        tuningSpecs.namedSequenceOpsToMove.push_back(namedSequenceOp);
+        continue;
+      }
+
+      // Extract the single ForeachMatchOp inside namedSequenceOp.
```

Comment:
I don't think this comments adds much value. Focus on documenting the __why__ instead of the __how__.